/hold cancel

Works as intended:

```
ERROR: pkg/controller/garbagecollector/garbagecollector.go:363:11: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 			logger.V(5).Error(err, "error syncing item", "item", n.identity)
ERROR: 			       ^
ERROR: pkg/controller/garbagecollector/graph_builder.go:225:10: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 		logger.V(4).Error(err, "unable to use a shared informer", "resource", resource, "kind", kind)
ERROR: 		       ^
ERROR: pkg/controller/resourcequota/resource_quota_monitor.go:178:9: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 	logger.V(4).Error(err, "QuotaMonitor unable to use a shared informer", "resource", resource.String())
ERROR: 	       ^
ERROR: pkg/controller/storageversionmigrator/storageversionmigrator.go:227:10: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 		logger.V(4).Error(err, "resource does not exist in our rest mapper", "gvr", gvr.String())
ERROR: 		       ^
ERROR: pkg/controller/storageversionmigrator/storageversionmigrator.go:244:10: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 		logger.V(4).Error(errMonitor, "resource does not exist in GC", "gvr", gvr.String())
ERROR: 		       ^
ERROR: pkg/controller/storageversionmigrator/storageversionmigrator.go:311:11: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 			logger.V(4).Error(err, "Unable to compare the resource version of the resource", "namespace", accessor.GetNamespace(), "name", accessor.GetName(), "gvr", gvr.String(), "accessorRV", accessor.GetResourceVersion(), "listResourceVersion", listResourceVersion, "error", err.Error())
ERROR: 			       ^
ERROR: pkg/controller/storageversionmigrator/storageversionmigrator.go:361:11: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 			logger.V(4).Error(errPatch, "Failed to migrate the resource", "namespace", accessor.GetNamespace(), "name", accessor.GetName(), "gvr", gvr.String(), "reason", apierrors.ReasonForError(errPatch))
ERROR: 			       ^
ERROR: pkg/controller/volume/selinuxwarning/selinux_warning_controller.go:476:11: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 			logger.V(4).Error(err, "failed to get SELinux label", "pod", klog.KObj(pod), "volume", mount)
ERROR: 			       ^
ERROR: pkg/controller/volume/selinuxwarning/selinux_warning_controller.go:539:11: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 			logger.V(2).Error(err, "failed to get first pod for event", "pod", conflict.Pod)
ERROR: 			       ^
ERROR: pkg/kubelet/allocation/allocation_manager.go:576:10: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 		logger.V(3).Error(err, "Failed to delete pod allocation", "podUID", uid)
ERROR: 		       ^
ERROR: pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:107:11: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 			logger.V(2).Error(err, "Failed to close grpc connection", "resource", c.Resource())
ERROR: 			       ^
ERROR: pkg/kubelet/cm/dra/plugin/dra_plugin_manager.go:398:13: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 					logger.V(3).Error(err, "Failed to establish WatchResources stream, will retry")
ERROR: 					       ^
ERROR: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:430:11: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 			logger.V(5).Error(fmt.Errorf("failed to parse /proc/self/cgroup: %w", err), warn)
ERROR: 			       ^
ERROR: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:441:11: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 			logger.V(5).Error(err, warn)
ERROR: 			       ^
ERROR: pkg/kubelet/lifecycle/handlers.go:85:11: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 			logger.V(1).Error(err, "Exec lifecycle hook for Container in Pod failed", "execCommand", handler.Exec.Command, "containerName", container.Name, "pod", klog.KObj(pod), "message", string(output))
ERROR: 			       ^
ERROR: pkg/kubelet/lifecycle/handlers.go:93:11: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 			logger.V(1).Error(err, "HTTP lifecycle hook for Container in Pod failed", "path", handler.HTTPGet.Path, "containerName", container.Name, "pod", klog.KObj(pod))
ERROR: 			       ^
ERROR: pkg/kubelet/lifecycle/handlers.go:101:11: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 			logger.V(1).Error(err, "Sleep lifecycle hook for Container in Pod failed", "sleepSeconds", handler.Sleep.Seconds, "containerName", container.Name, "pod", klog.KObj(pod))
ERROR: 			       ^
ERROR: pkg/kubelet/lifecycle/handlers.go:152:10: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 		logger.V(1).Error(err, "HTTPS request to lifecycle hook got HTTP response, retrying with HTTP.", "pod", klog.KObj(pod), "host", req.URL.Host)
ERROR: 		       ^
ERROR: pkg/kubelet/metrics/collectors/cri_metrics.go:110:10: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 		logger.V(5).Error(err, "Descriptor not present in pre-populated list of descriptors", "name", m.Name)
ERROR: 		       ^
ERROR: pkg/kubelet/pleg/generic.go:300:13: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 			g.logger.V(4).Error(err, "PLEG: Ignoring events for pod", "pod", klog.KRef(pod.Namespace, pod.Name))
ERROR: 			         ^
ERROR: pkg/kubelet/prober/prober.go:106:10: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 		logger.V(1).Error(err, "Probe errored", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "probeResult", result)
ERROR: 		       ^
ERROR: pkg/kubelet/prober/prober_manager.go:197:12: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 				logger.V(8).Error(nil, "Startup probe already exists for container",
ERROR: 				       ^
ERROR: pkg/kubelet/prober/prober_manager.go:209:12: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 				logger.V(8).Error(nil, "Readiness probe already exists for container",
ERROR: 				       ^
ERROR: pkg/kubelet/prober/prober_manager.go:221:12: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 				logger.V(8).Error(nil, "Liveness probe already exists for container",
ERROR: 				       ^
ERROR: pkg/kubelet/stats/cri_stats_provider.go:168:10: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 		logger.V(5).Error(err,
ERROR: 		       ^
ERROR: pkg/kubelet/stats/helper.go:502:10: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 		logger.V(6).Error(err, "Unable to fetch pod log stats", "pod", klog.KRef(podNs, podName))
ERROR: 		       ^
ERROR: pkg/kubelet/stats/helper.go:510:10: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 		logger.V(6).Error(err, "Unable to fetch pod etc hosts stats", "pod", klog.KRef(podNs, podName))
ERROR: 		       ^
ERROR: pkg/kubelet/volumemanager/cache/desired_state_of_world.go:673:9: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 	logger.V(4).Error(err, "Please report this error in https://github.com/kubernetes/enhancements/issues/1710, together with full Pod yaml file")
ERROR: 	       ^
ERROR: pkg/kubelet/volumemanager/cache/desired_state_of_world.go:688:10: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 		logger.V(4).Error(err, "failed to get CSI driver name from volume spec")
ERROR: 		       ^
ERROR: pkg/kubelet/volumemanager/reconciler/reconstruct.go:188:10: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 		logger.V(4).Error(fetchErr, "Failed to get Node status to reconstruct device paths")
ERROR: 		       ^
ERROR: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go:418:25: V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err). (logcheck)
ERROR: 		klog.FromContext(ctx).V(1).Error(err, "failed to authorize")
ERROR: 		                      ^
31 issues:
* logcheck: 31
```

--- a/logcheck/README.md
+++ b/logcheck/README.md
@@ -105,6 +105,20 @@ runtime in klog.
 
 This check flags all invocation of `klog.V(0)` or any of it's equivalent as errors
 
+## verbosity-error (enabled by default)
+
+`logger.V(5).Error` for a `logr.Logger` instance is identical to `logger.Error`
+because `logger.V(5)` is just another `Logger` instance and `Error` bypasses
+all verbosity checks. This is different from `klog.V(5).ErrorS`, which does
+the verbosity check.
+
+If the call really is an `Error` log call, i.e. with "an admin must know about
+this" importance, then the preceding `V()` should be removed. If the semantic
+from klog is desired, then the corresponding go-logr call is
+`logger.V(5).Info`, with the error passed in a `"err", error` key/value
+pair - at least in Kubernetes. Other projects may have different naming
+conventions.
+
 ## key (enabled by default)
 
 This check flags check whether name arguments are valid keys according to the

--- a/logcheck/main_test.go
+++ b/logcheck/main_test.go
@@ -112,17 +112,19 @@ func TestAnalyzer(t *testing.T) {
 		{
 			name: "Do not allow Verbosity Zero logs",
 			enabled: map[string]string{
-				"structured": "false",
-				"key":        "false",
+				"structured":      "false",
+				"key":             "false",
+				"verbosity-error": "false",
 			},
 			testPackage: "doNotAllowVerbosityZeroLogs",
 		},
 		{
 			name: "Allow Verbosity Zero logs",
 			enabled: map[string]string{
-				"structured":     "false",
-				"verbosity-zero": "false",
-				"key":            "false",
+				"structured":      "false",
+				"verbosity-zero":  "false",
+				"verbosity-error": "false",
+				"key":             "false",
 			},
 			testPackage: "allowVerbosityZeroLogs",
 		},

--- a/logcheck/pkg/logcheck.go
+++ b/logcheck/pkg/logcheck.go
@@ -34,14 +34,15 @@ import (
 )
 
 const (
-	structuredCheck    = "structured"
-	parametersCheck    = "parameters"
-	contextualCheck    = "contextual"
-	withHelpersCheck   = "with-helpers"
-	verbosityZeroCheck = "verbosity-zero"
-	keyCheck           = "key"
-	valueCheck         = "value"
-	deprecationsCheck  = "deprecations"
+	structuredCheck     = "structured"
+	parametersCheck     = "parameters"
+	contextualCheck     = "contextual"
+	withHelpersCheck    = "with-helpers"
+	verbosityZeroCheck  = "verbosity-zero"
+	verbosityErrorCheck = "verbosity-error"
+	keyCheck            = "key"
+	valueCheck          = "value"
+	deprecationsCheck   = "deprecations"
 )
 
 type checks map[string]*bool
@@ -72,14 +73,15 @@ func (c *Config) ParseConfig(configContent string) error {
 func Analyser() (*analysis.Analyzer, *Config) {
 	c := Config{
 		enabled: checks{
-			structuredCheck:    new(bool),
-			parametersCheck:    new(bool),
-			contextualCheck:    new(bool),
-			withHelpersCheck:   new(bool),
-			verbosityZeroCheck: new(bool),
-			keyCheck:           new(bool),
-			valueCheck:         new(bool),
-			deprecationsCheck:  new(bool),
+			structuredCheck:     new(bool),
+			parametersCheck:     new(bool),
+			contextualCheck:     new(bool),
+			withHelpersCheck:    new(bool),
+			verbosityZeroCheck:  new(bool),
+			verbosityErrorCheck: new(bool),
+			keyCheck:            new(bool),
+			valueCheck:          new(bool),
+			deprecationsCheck:   new(bool),
 		},
 	}
 	c.fileOverrides.validChecks = map[string]bool{}
@@ -94,6 +96,7 @@ klog methods (Info, Infof, Error, Errorf, Warningf, etc).`)
 	logcheckFlags.BoolVar(c.enabled[contextualCheck], prefix+contextualCheck, false, `When true, logcheck will only allow log calls for contextual logging (retrieving a Logger from klog or the context and logging through that) and warn about all others.`)
 	logcheckFlags.BoolVar(c.enabled[withHelpersCheck], prefix+withHelpersCheck, false, `When true, logcheck will warn about direct calls to WithName, WithValues and NewContext.`)
 	logcheckFlags.BoolVar(c.enabled[verbosityZeroCheck], prefix+verbosityZeroCheck, true, `When true, logcheck will check whether the parameter for V() is 0.`)
+	logcheckFlags.BoolVar(c.enabled[verbosityErrorCheck], prefix+verbosityErrorCheck, true, `When true, logcheck will check for V() in front of a logr Error call.`)
 	logcheckFlags.BoolVar(c.enabled[keyCheck], prefix+keyCheck, true, `When true, logcheck will check whether name arguments are valid keys according to the guidelines in (https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#name-arguments).`)
 	logcheckFlags.BoolVar(c.enabled[valueCheck], prefix+valueCheck, false, `When true, logcheck will check for problematic values (for example, types that have an incomplete fmt.Stringer implementation).`)
 	logcheckFlags.BoolVar(c.enabled[deprecationsCheck], prefix+deprecationsCheck, true, `When true, logcheck will analyze the usage of deprecated Klog function calls.`)
@@ -292,6 +295,18 @@ func checkForFunctionExpr(fexpr *ast.CallExpr, pass *analysis.Pass, c *Config) {
 			// verbosity Zero Check
 			if c.isEnabled(verbosityZeroCheck, filename) {
 				checkForVerbosityZero(fexpr, pass)
+			}
+
+			if fName == "Error" && c.isEnabled(verbosityErrorCheck, filename) {
+				// Is X itself the result of a selector expression with V as method name?
+				if innerCallExpr, ok := selExpr.X.(*ast.CallExpr); ok {
+					if innerSelExpr, ok := innerCallExpr.Fun.(*ast.SelectorExpr); ok && innerSelExpr.Sel.Name == "V" {
+						pass.Report(analysis.Diagnostic{
+							Pos:     innerSelExpr.Sel.Pos(),
+							Message: `V().Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V().Info(..., "err", err).`,
+						})
+					}
+				}
 			}
 		} else if fName == "NewContext" &&
 			isPackage(selExpr.X, "github.com/go-logr/logr", pass) &&

--- a/logcheck/testdata/src/gologr/gologr.go
+++ b/logcheck/testdata/src/gologr/gologr.go
@@ -33,7 +33,7 @@ func logging() {
 	logger.WithValues("missing value")    // want `Additional arguments to WithValues should always be Key Value pairs. Please check if there is any key or value missing.`
 
 	logger.V(1).Info("hello", "missing value") // want `Additional arguments to Info should always be Key Value pairs. Please check if there is any key or value missing.`
-	logger.V(1).Error(nil, "hello", 1, 2)      // want `Key positional arguments are expected to be inlined constant strings. Please replace 1 provided with string value`
+	logger.V(1).Error(nil, "hello", 1, 2)      // want `Key positional arguments are expected to be inlined constant strings. Please replace 1 provided with string value` `V\(\).Error ignores the verbosity and always logs. Use only Error if that is desired, otherwise V\(\).Info\(..., "err", err\).`
 	logger.V(1).WithValues("missing value")    // want `Additional arguments to WithValues should always be Key Value pairs. Please check if there is any key or value missing.`
 
 	// variadic input to logger.Info, logger.Error, logger.WithValues functions
@@ -43,4 +43,5 @@ func logging() {
 	logger.WithValues(kvs...)
 	logger.WithValues(kvs)                      // want `Additional arguments to WithValues should always be Key Value pairs. Please check if there is any key or value missing.`
 	logger.Error(nil, "foo error message", kvs) // want `Additional arguments to Error should always be Key Value pairs. Please check if there is any key or value missing.`
+
 }

